scripts/criu-ns

```diff
@@ -18,190 +18,238 @@ MS_SLAVE = 1 << 19
 _libc = ctypes.CDLL(ctypes.util.find_library("c"), use_errno=True)
 
 try:
-	_unshare = _libc.unshare
+    _unshare = _libc.unshare
 except AttributeError:
-	raise OSError(errno.EINVAL, "unshare is not supported on this platform")
+    raise OSError(errno.EINVAL, "unshare is not supported on this platform")
 else:
-	_unshare.argtypes = [ ctypes.c_int ]
-	_unshare.restype = ctypes.c_int
+    _unshare.argtypes = [ctypes.c_int]
+    _unshare.restype = ctypes.c_int
 
 try:
-	_setns = _libc.setns
+    _setns = _libc.setns
 except AttributeError:
-	raise OSError(errno.EINVAL, "setns is not supported on this platform")
+    raise OSError(errno.EINVAL, "setns is not supported on this platform")
 else:
-	_setns.argtypes = [ ctypes.c_int, ctypes.c_int ]
-	_setns.restype = ctypes.c_int
+    _setns.argtypes = [ctypes.c_int, ctypes.c_int]
+    _setns.restype = ctypes.c_int
 
 try:
-	_mount = _libc.mount
+    _mount = _libc.mount
 except AttributeError:
-	raise OSError(errno.EINVAL, "mount is not supported on this platform")
+    raise OSError(errno.EINVAL, "mount is not supported on this platform")
 else:
-	_mount.argtypes = [
-		ctypes.c_char_p,
-		ctypes.c_char_p,
-		ctypes.c_char_p,
-		ctypes.c_ulong,
-		ctypes.c_void_p
-	]
-	_mount.restype = ctypes.c_int
+    _mount.argtypes = [
+        ctypes.c_char_p,
+        ctypes.c_char_p,
+        ctypes.c_char_p,
+        ctypes.c_ulong,
+        ctypes.c_void_p
+    ]
+    _mount.restype = ctypes.c_int
 
 try:
-	_umount = _libc.umount
+    _umount = _libc.umount
 except AttributeError:
-	raise OSError(errno.EINVAL, "umount is not supported on this platform")
+    raise OSError(errno.EINVAL, "umount is not supported on this platform")
 else:
-	_umount.argtypes = [ctypes.c_char]
-	_umount.restype = ctypes.c_int
+    _umount.argtypes = [ctypes.c_char]
+    _umount.restype = ctypes.c_int
 
 
 def run_criu():
-	print(sys.argv)
-	os.execlp('criu', *['criu'] + sys.argv[1:])
+    print(sys.argv)
+    os.execlp('criu', *['criu'] + sys.argv[1:])
+
+
+def _mount_new_proc():
+    """
+    Mount new /proc filesystem.
+    """
+    if _mount(None, b"/", None, MS_SLAVE|MS_REC, None):
```

Comment:
Why not MS_PRIVATE? why do you need MS_SLAVE, for which exact propagation?

Reply:
I'm not strictly against it as it's code copying, but this propagation is useless FMPOV.

Reply:
If we create a mount namespace just for mounting /proc, MS_SLAVE is the best choice. All changes in the host mount namespace will be propagated into this namespace.

Reply:
> If we create a mount namespace just for mounting /proc, MS_SLAVE is the best choice. All changes in the host mount namespace will be propagated into this namespace.

*offtop*

@avagin Note for the case of "umount a file system and detach its device from the system" probably a more robust way would be not to rely on propagation but to create some new syscall umount3 (or smth. like this) to just umount all mounts of the superblock.

It would probably also help with autofs problems: now when autofs service is restarted it tries to umount autofs and mount a new one in it's place, if you clone mountns and make it rprivate, bindmount of old autofs mount in it will remain but be in some kind of broken state. This new syscall can give an ability to umount all kinds of leftover mounts.